src/useMutationObserver.ts

```diff
@@ -0,0 +1,68 @@
+import useCustomEffect, { IsEqual } from './useCustomEffect'
+import isEqual from 'lodash/isEqual'
+import useImmediateUpdateEffect from './useImmediateUpdateEffect'
+import useMountEffect from './useMountEffect'
+import useEventCallback from './useEventCallback'
+import { useRef } from 'react'
+
+type Deps = [Element | null | undefined, MutationObserverInit]
+
+const isDepsEqual: IsEqual<Deps> = (prev, next) =>
+  prev[0] === next[0] && isEqual(prev[1], next[1])
+
+/**
+ * Observe mutations on a DOM node or tree of DOM nodes.
+ * Depends on the `MutationObserver` api.
+ *
+ * ```ts
+ * const [element, attachRef] = useCallbackRef(null);
+ *
+ * useMutationObserver(element, { subtree: true }, (records) => {
+ *
+ * });
+ *
+ * return (
+ *   <div ref={attachRef} />
+ * )
+ * ```
+ *
+ * @param element The DOM element to observe
+ * @param config The observer configuration
+ * @param callback A callback fired when a mutation occurs
+ */
+function useMutationObserver(
+  element: Element | null | undefined,
+  config: MutationObserverInit,
+  callback: MutationCallback,
+): void {
+  const observerRef = useRef<MutationObserver | null>()
+  const fn = useEventCallback(callback)
+
+  useMountEffect(() => {
+    if (!element) return
+
+    observerRef.current = new MutationObserver(fn)
```

Comment:
```suggestion
    observerRef.current = new MutationObserver(fn)
    observer.observe(element, config)
```

don't you need this here? not that there can really be an element on immediate mount